apps/condo/domains/user/schema/ConfirmPhone.test.js

```diff
@@ -0,0 +1,143 @@
+    query findConfirmPhoneFromToken($token: String!) {
+        confirmPhoneActions: allConfirmPhoneActions(where: { token: $token }) {
+            id
+            phone
+            smsCode
+            isPhoneVerified
+            smsCodeRequestedAt
+            smsCodeExpiresAt
+            requestedAt
+            expiresAt
+            retries
+        }
+    }
+`
+
+const UPDATE_CONFIRM_PHONE_ACTION = gql`
+    mutation updateConfirmPhoneAction ($id: ID!, $data: ConfirmPhoneActionUpdateInput) {
+        result: updateConfirmPhoneAction(id: $id, data: $data) {
+            id            
+        }
+    }
+`
+
+// TODO(zuch): Discuss is it possible to turn off real sms sending during tests
+
+describe('ConfirmPhoneAction Service', () => {
+    it('can be created by Anonymous', async () => {
+        const client = await makeClient()       
+        const phone = createTestPhone()
+        const { data: { token } } = await client.mutate(START_CONFIRM_PHONE_MUTATION, { phone, dv: 1, sender: { dv: 1, fingerprint: 'tests' } })
+        expect(token).not.toHaveLength(0)
+    })
+
+    it('throw error when is confirming with wrong sms code', async () => {
+        const client = await makeClient()       
+        const phone = createTestPhone()
+        const { data: { token } }  = await client.mutate(START_CONFIRM_PHONE_MUTATION, { phone, dv: 1, sender: { dv: 1, fingerprint: 'tests' } })
+        const wrongLengthSmsCode = 11111
+        const res = await client.mutate(COMPLETE_CONFIRM_PHONE_MUTATION, { token, smsCode: wrongLengthSmsCode })
+        expect(JSON.stringify(res.errors)).toContain(CONFIRM_PHONE_SMS_CODE_VERIFICATION_FAILED)
+    })
+
+    it('should increment retries on failed attempt', async () => {
+        const client = await makeClient()       
+        const phone = createTestPhone()
+        const { data: { token } }  = await client.mutate(START_CONFIRM_PHONE_MUTATION, { phone, dv: 1, sender: { dv: 1, fingerprint: 'tests' } })
+        const wrongLengthSmsCode = 11111
+        await client.mutate(COMPLETE_CONFIRM_PHONE_MUTATION, { token, smsCode: wrongLengthSmsCode })
+        const admin = await makeLoggedInAdminClient()
+        const { data: { confirmPhoneActions: [actionAfter] } } = await admin.query(GET_CONFIRM_PHONE_BY_TOKEN, { token })
+        expect(actionAfter.retries).toBe(1)        
+    })
+
+    it('marks itself as verified when sms code matches', async () => {
+        const client = await makeClient()       
+        const admin = await makeLoggedInAdminClient()        
+        const phone = createTestPhone()
+        const { data: { token } }  = await client.mutate(START_CONFIRM_PHONE_MUTATION, { phone, dv: 1, sender: { dv: 1, fingerprint: 'tests' } })
+        const { data: { confirmPhoneActions: [actionBefore] } } = await admin.query(GET_CONFIRM_PHONE_BY_TOKEN, { token })
+        const { data: { status } } = await client.mutate(COMPLETE_CONFIRM_PHONE_MUTATION, { token, smsCode: actionBefore.smsCode })
```

Comment:
is it ok, that we can get smsCode from GET_CONFIRM_PHONE_BY_TOKEN query?

Reply:
This query is used only in tests. There is a query for the user and it's located in ConfirmPhone.js
```js
 const { errors, data } = await context.executeGraphQL({
                    context: context.createContext({ skipAccessControl: true }),
                    query: `
                        query findConfirmPhoneFromToken($token: String!, $now: String!) {
                          confirmPhoneActions: allConfirmPhoneActions(where: { token: $token, expiresAt_gte: $now, completedAt: null }) {
                            id
                            phone
                          }
                        }
                    `,
                    variables: { token, now: new Date(now).toISOString() },
                })
```